web-admin/src/features/dashboards/query-mappers/getDashboardFromAggregationRequest.ts

```diff
@@ -0,0 +1,98 @@
+import {
+  SortDirection,
+  SortType,
+} from "@rilldata/web-common/features/dashboards/proto-state/derived-types";
+import { createAndExpression } from "@rilldata/web-common/features/dashboards/stores/filter-utils";
+import { TDDChart } from "@rilldata/web-common/features/dashboards/time-dimension-details/types";
+import { DashboardState_ActivePage } from "@rilldata/web-common/proto/gen/rill/ui/v1/dashboard_pb";
+import type { V1MetricsViewAggregationRequest } from "@rilldata/web-common/runtime-client";
+
+export async function getDashboardFromAggregationRequest({
+  queryClient,
+  instanceId,
+  req,
+  dashboard,
+  timeRangeSummary,
+  executionTime,
+  metricsView,
+}: QueryMapperArgs<V1MetricsViewAggregationRequest>) {
+  if (req.timeRange) {
+    dashboard.selectedTimeRange = getSelectedTimeRange(
+      req.timeRange,
+      timeRangeSummary,
+      req.timeRange.isoDuration ?? "",
+      executionTime,
+    );
+  }
+
+  if (req.where) dashboard.whereFilter = req.where;
+  if (req.having?.cond?.exprs?.length && req.dimensions?.[0]?.name) {
+    const dimension = req.dimensions[0].name;
+    if (req.having.cond.exprs.length > 1) {
+      const expr = await convertExprToToplist(
+        queryClient,
+        instanceId,
+        dashboard.name,
+        dimension,
+        req.measures?.[0]?.name ?? "",
+        dashboard.selectedTimeRange,
+        req.where,
+        req.having,
+      );
+      if (expr) {
+        dashboard.whereFilter = mergeFilters(
+          dashboard.whereFilter ?? createAndExpression([]),
+          createAndExpression([expr]),
+        );
+      }
+    } else {
+      dashboard.dimensionThresholdFilters = [
+        {
+          name: dimension,
+          filter: createAndExpression([req.having.cond?.exprs?.[0]]),
+        },
+      ];
+    }
+  }
+
+  if (req.timeRange?.timeZone) {
+    dashboard.selectedTimezone = req.timeRange?.timeZone || "UTC";
+  }
+
+  dashboard.visibleMeasureKeys = new Set(
+    req.measures?.map((m) => m.name ?? "") ?? [],
+  );
+
+  // if the selected sort is a measure set it to leaderboardMeasureName
+  if (
+    req.sort?.[0] &&
+    (metricsView.measures?.findIndex((m) => m.name === req.sort?.[0]?.name) ??
+      -1) >= 0
+  ) {
+    dashboard.leaderboardMeasureName = req.sort[0].name ?? "";
+    dashboard.sortDirection = req.sort[0].desc
+      ? SortDirection.DESCENDING
+      : SortDirection.ASCENDING;
+    dashboard.dashboardSortType = SortType.VALUE;
+  }
+
+  if (req.dimensions?.length) {
+    dashboard.selectedDimensionName = req.dimensions[0].name;
+    dashboard.activePage = DashboardState_ActivePage.DIMENSION_TABLE;
+  } else {
+    dashboard.tdd = {
+      chartType: TDDChart.DEFAULT,
+      expandedMeasureName: req.measures?.[0]?.name ?? "",
+      pinIndex: -1,
+    };
+    dashboard.activePage = DashboardState_ActivePage.TIME_DIMENSIONAL_DETAIL;
+  }
+
+  return dashboard;
+}
```

Comment:
**Issue**: The function 'getDashboardFromAggregationRequest' is complex and handles multiple responsibilities, which can make it difficult to maintain and test. It modifies the dashboard state based on various conditions and inputs, which increases the risk of bugs due to side effects. <br> **Fix**: Refactor the function to separate concerns more clearly. Break down the function into smaller, more focused functions, each handling a specific part of the dashboard state modification. This will improve maintainability and make unit testing easier. <br> **Code Suggestion**: 
 ```
 +import { updateDashboardTimeRange, updateDashboardFilters, updateDashboardSorting } from './dashboardStateHelpers';
 +
 +export async function getDashboardFromAggregationRequest({
 +  queryClient,
 +  instanceId,
 +  req,
 +  dashboard,
 +  timeRangeSummary,
 +  executionTime,
 +  metricsView,
 +}: QueryMapperArgs<V1MetricsViewAggregationRequest>) {
 +  updateDashboardTimeRange(req, dashboard, timeRangeSummary, executionTime);
 +  updateDashboardFilters(req, dashboard, queryClient, instanceId);
 +  updateDashboardSorting(req, dashboard, metricsView);
 +}
 ```

Reply:
IMO this is overkill, it is better to refactor once we have other types of queries and we can split out common code.